File: src/main/java/com/qsoftware/bubbles/core/GlobalExceptionHandler.java
package com.qsoftware.bubbles.core;

import com.qsoftware.bubbles.QBubbles;
import org.apache.logging.log4j.Logger;

import java.io.PrintStream;
import java.io.PrintWriter;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.Set;

public class GlobalExceptionHandler implements Thread.UncaughtExceptionHandler {
    private static Logger logger = QBubbles.getLogger();

    static {
//        try {
//            logger = new GameLogger("STDERR");
//        } catch (InstanceAlreadyExistsException e) {
//            e.printStackTrace();
//        }
    }

    private static boolean hooked = false;

    public void uncaughtException(Thread t, Throwable e) {
//        logger.fatal("Unhandled exception caught!");
        printStackTrace(e);
        System.exit(1);
    }

    /**
     * Message for trying to suppress a null exception.
     */
    private static final String NULL_CAUSE_MESSAGE = "Cannot suppress a null exception.";

    /**
     * Message for trying to suppress oneself.
     */
    private static final String SELF_SUPPRESSION_MESSAGE = "Self-suppression not permitted";

    /**
     * Caption  for labeling causative exception stack traces
     */
    private static final String CAUSE_CAPTION = "Caused by: ";

    /**
     * Caption for labeling suppressed exception stack traces
     */
    private static final String SUPPRESSED_CAPTION = "Suppressed: ";

    private void printStackTrace(Throwable e) {
        PrintStreamOrWriter s = new WrappedPrintStream(System.err);


        // Guard against malicious overrides of Throwable.equals by
        // using a Set with identity equality semantics.
        Set<Throwable> dejaVu = Collections.newSetFromMap(new IdentityHashMap<>());
        dejaVu.add(e);

        synchronized (s.lock()) {
            // Print our stack trace
            logger.fatal(e.toString());
            StackTraceElement[] trace = e.getStackTrace();
            for (StackTraceElement traceElement : trace)
                logger.fatal("\tat " + traceElement);

            // Print suppressed exceptions, if any
            for (Throwable se : e.getSuppressed())
                printEnclosedStackTrace(se, s, trace, SUPPRESSED_CAPTION, "\t", dejaVu);

            // Print cause, if any
            Throwable ourCause = e.getCause();
            if (ourCause != null)
                printEnclosedStackTrace(ourCause, s, trace, CAUSE_CAPTION, "", dejaVu);
        }
    }

    /**
     * Print our stack trace as an enclosed exception for the specified
     * stack trace.
     */
    private void printEnclosedStackTrace(Throwable e, PrintStreamOrWriter s,
                                         StackTraceElement[] enclosingTrace,
                                         String caption,
                                         String prefix,
                                         Set<Throwable> dejaVu) {
        assert Thread.holdsLock(s.lock());
        if (dejaVu.contains(e)) {
            s.println(prefix + caption + "[CIRCULAR REFERENCE: " + this + "]");
        } else {
            dejaVu.add(e);
            // Compute number of frames in common between this and enclosing trace
            StackTraceElement[] trace = e.getStackTrace();
            int m = trace.length - 1;
            int n = enclosingTrace.length - 1;
            while (m >= 0 && n >= 0 && trace[m].equals(enclosingTrace[n])) {
                m--;
                n--;
            }
            int framesInCommon = trace.length - 1 - m;

            // Print our stack trace
            s.println(prefix + caption + this);
            for (int i = 0; i <= m; i++)
                s.println(prefix + "\tat " + trace[i]);
            if (framesInCommon != 0)
                s.println(prefix + "\t... " + framesInCommon + " more");

            // Print suppressed exceptions, if any
            for (Throwable se : e.getSuppressed())
                printEnclosedStackTrace(se, s, trace, SUPPRESSED_CAPTION,
                        prefix + "\t", dejaVu);

            // Print cause, if any
            Throwable ourCause = e.getCause();
            if (ourCause != null)
                printEnclosedStackTrace(ourCause, s, trace, CAUSE_CAPTION, prefix, dejaVu);
        }
    }

    public static void setLogger(Logger logger) throws IllegalStateException {
        if (hooked) throw new IllegalStateException("Not allowed to access after hooked!");
        hooked = true;
        GlobalExceptionHandler.logger = logger;
    }

    /**
     * Wrapper class for PrintStream and PrintWriter to enable a single
     * implementation of printStackTrace.
     */
    private abstract static class PrintStreamOrWriter {
        /**
         * Returns the object to be locked when using this StreamOrWriter
         */
        abstract Object lock();

        /**
         * Prints the specified string as a line on this StreamOrWriter
         */
        abstract void println(Object o);
    }

    private static class WrappedPrintStream extends PrintStreamOrWriter {
        private final PrintStream printStream;

        WrappedPrintStream(PrintStream printStream) {
            this.printStream = printStream;
        }

        Object lock() {
            return printStream;
        }

        void println(Object o) {
            logger.fatal(o.toString());
        }
    }

    private static class WrappedPrintWriter extends PrintStreamOrWriter {
        private final PrintWriter printWriter;

        WrappedPrintWriter(PrintWriter printWriter) {
            this.printWriter = printWriter;
        }

        Object lock() {
            return printWriter;
        }

        void println(Object o) {
            logger.fatal(o.toString());
        }
    }
}

File: src/main/kotlin/com/qtech/bubbles/core/GlobalExceptionHandler.kt
package com.qtech.bubbles.core

import com.qtech.bubbles.BubbleBlaster
import org.apache.logging.log4j.Logger
import java.io.PrintStream
import java.io.PrintWriter
import java.util.*

class GlobalExceptionHandler : Thread.UncaughtExceptionHandler {
    override fun uncaughtException(t: Thread, e: Throwable) {
//        logger.fatal("Unhandled exception caught!");
        printStackTrace(e)
        System.exit(1)
    }

    private fun printStackTrace(e: Throwable) {
        val s: PrintStreamOrWriter = WrappedPrintStream(System.err)


        // Guard against malicious overrides of Throwable.equals by
        // using a Set with identity equality semantics.
        val dejaVu = Collections.newSetFromMap(IdentityHashMap<Throwable, Boolean>())
        dejaVu.add(e)
        synchronized(s.lock()) {

            // Print our stack trace
            logger.fatal(e.toString())
            val trace = e.stackTrace
            for (traceElement in trace) logger.fatal("\tat $traceElement")

            // Print suppressed exceptions, if any
            for (se in e.suppressed) printEnclosedStackTrace(se, s, trace, SUPPRESSED_CAPTION, "\t", dejaVu)

            // Print cause, if any
            val ourCause = e.cause
            if (ourCause != null) printEnclosedStackTrace(ourCause, s, trace, CAUSE_CAPTION, "", dejaVu)
        }
    }

    /**
     * Print our stack trace as an enclosed exception for the specified
     * stack trace.
     */
    private fun printEnclosedStackTrace(
        e: Throwable, s: PrintStreamOrWriter,
        enclosingTrace: Array<StackTraceElement>,
        caption: String,
        prefix: String,
        dejaVu: MutableSet<Throwable>
    ) {
        assert(Thread.holdsLock(s.lock()))
        if (dejaVu.contains(e)) {
            s.println("$prefix$caption[CIRCULAR REFERENCE: $this]")
        } else {
            dejaVu.add(e)
            // Compute number of frames in common between this and enclosing trace
            val trace = e.stackTrace
            var m = trace.size - 1
            var n = enclosingTrace.size - 1
            while (m >= 0 && n >= 0 && trace[m] == enclosingTrace[n]) {
                m--
                n--
            }
            val framesInCommon = trace.size - 1 - m

            // Print our stack trace
            s.println(prefix + caption + this)
            for (i in 0..m) s.println(prefix + "\tat " + trace[i])
            if (framesInCommon != 0) s.println("$prefix\t... $framesInCommon more")

            // Print suppressed exceptions, if any
            for (se in e.suppressed) printEnclosedStackTrace(
                se, s, trace, SUPPRESSED_CAPTION,
                prefix + "\t", dejaVu
            )

            // Print cause, if any
            val ourCause = e.cause
            if (ourCause != null) printEnclosedStackTrace(ourCause, s, trace, CAUSE_CAPTION, prefix, dejaVu)
        }
    }

    /**
     * Wrapper class for PrintStream and PrintWriter to enable a single
     * implementation of printStackTrace.
     */
    private abstract class PrintStreamOrWriter {
        /**
         * Returns the object to be locked when using this StreamOrWriter
         */
        abstract fun lock(): Any

        /**
         * Prints the specified string as a line on this StreamOrWriter
         */
        abstract fun println(o: Any)
    }

    private class WrappedPrintStream(private val printStream: PrintStream) : PrintStreamOrWriter() {
        override fun lock(): Any {
            return printStream
        }

        override fun println(o: Any) {
            logger.fatal(o.toString())
        }
    }

    private class WrappedPrintWriter(private val printWriter: PrintWriter) : PrintStreamOrWriter() {
        override fun lock(): Any {
            return printWriter
        }

        override fun println(o: Any) {
            logger.fatal(o.toString())
        }
    }

    companion object {
        private var logger = BubbleBlaster.logger
        private var hooked = false

        /**
         * Message for trying to suppress a null exception.
         */
        private const val NULL_CAUSE_MESSAGE = "Cannot suppress a null exception."

        /**
         * Message for trying to suppress oneself.
         */
        private const val SELF_SUPPRESSION_MESSAGE = "Self-suppression not permitted"

        /**
         * Caption  for labeling causative exception stack traces
         */
        private const val CAUSE_CAPTION = "Caused by: "

        /**
         * Caption for labeling suppressed exception stack traces
         */
        private const val SUPPRESSED_CAPTION = "Suppressed: "
        @Throws(IllegalStateException::class)
        fun setLogger(logger: Logger) {
            check(!hooked) { "Not allowed to access after hooked!" }
            hooked = true
            Companion.logger = logger
        }
    }
}